The project will analyze EBay car sales.

In [24]:
import numpy as np
import pandas as pd

autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [25]:
# Exploring autos #

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [26]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [27]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Most of the data seem to be strings - I will need to format the data differently, so that it can be analyzed.

# As a next step, I will work on the column names.

In [28]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [29]:
# creating mapping #

mapping = {"yearOfRegistration": "registration_year",
           "monthOfRegistration": "registration_month",
           "notRepairedDamage": "unrepaired_damage",
           "dateCreated": "ad_created"
          }

autos.rename(columns=mapping, inplace=True)
print(autos.columns)
print(autos.head())

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2 

Looking at the values for all columns.

In [30]:
autos.describe(include='all')

dateCrawled         name  seller offerType  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicleType  registration_year  gearbox       powerPS  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuelType       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrOfPictures    postalCode             lastSeen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

We can see that for a number of columns, there are only 2 possible values, with 1 of them taking the majority of values - these columns are: 
- **seller**;
- **offerType**

**nrofPictures** should also be excluded, because it is an integer, but all of its values are zeros.

Examining price and odometer year in more detail:

In [31]:
print(autos["price"].head())
print(autos["odometer"].head())

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object
0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object


Converting both columns to integers

In [32]:
autos["price"] = autos["price"].str.replace("$","").str.replace(",","").astype(int)
print(autos["price"].head())
autos["odometer"] = autos["odometer"].str.replace(",","").str.replace("km","").astype(int)
print(autos["odometer"].head())

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64
0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer, dtype: int64


In [33]:
autos.rename(columns={"odometer":"odometer_km"},inplace=True)

In [34]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer_km', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [35]:
# exploring the price and odometer_km columns #

print(autos["price"].unique())
print("\n")
print(autos["price"].value_counts())
print("\n")
print(autos["odometer_km"].unique())
print("\n")
print(autos["odometer_km"].value_counts())

[ 5000  8500  8990 ...   385 22200 16995]


0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64


[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]


150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64


# Examine outliers in the price column

In [36]:
print(autos["price"].value_counts().head())
print("\n")
print(autos["price"].value_counts().sort_index(ascending=True))

0       1421
500      781
1500     734
2500     643
1000     639
Name: price, dtype: int64


0           1421
1            156
2              3
3              1
5              2
            ... 
10000000       1
11111111       2
12345678       3
27322222       1
99999999       1
Name: price, Length: 2357, dtype: int64


Some of the prices do not seem to be realistic. There are 1,421 cars with 0 price, and some cars have prices in the millions.

I will remove all cars with values with values higher than 10000000 USD as there are very few such vehicles and also vehicles with price of $0 (something sold/bought should always have a value). The issue with removing $0 vehicles is that it is the most frequent category.

In [37]:
autos = autos[autos["price"].between(1, 9999999)]
print(autos["price"].value_counts().sort_index(ascending=True))

1          156
2            3
3            1
5            2
8            1
          ... 
999990       1
999999       2
1234566      1
1300000      1
3890000      1
Name: price, Length: 2351, dtype: int64


# Examine outliers in the odometer_km column

In [38]:
print(autos["odometer_km"].value_counts().head())
print("\n")
print(autos["odometer_km"].value_counts().sort_index(ascending=True))

150000    31417
125000     5058
100000     2115
90000      1734
80000      1415
Name: odometer_km, dtype: int64


5000        837
10000       253
20000       762
30000       780
40000       815
50000      1013
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5058
150000    31417
Name: odometer_km, dtype: int64


I would not remove any outliers here, although it is interesting that the odometer max values are 150,000 km - I would have expected that there are cars with higher odometer readings on sale as well.

In [39]:
print(autos.head())

           dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType  price   abtest vehicleType  registration_year    gearbox  \
0  privat   Angebot   5000  control         bus               2004    manuell   
1  privat   Angebot   8500  control   limousine               1997  automatik   
2  privat   Angebot   8990     test   limousine               2009    manuell   
3  privat   Angebot   4350  control  kleinwagen               2007  automatik   
4  privat   Angebot   1350     test       kombi               2003    manuell   

   powerPS   model  odometer_km  registration_mont

# Working with dates

In [40]:
autos[["dateCrawled","ad_created","lastSeen"]]

dateCrawled           ad_created             lastSeen
0      2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1      2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2      2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3      2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4      2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
...                    ...                  ...                  ...
49995  2016-03-27 14:38:19  2016-03-27 00:00:00  2016-04-01 13:47:40
49996  2016-03-28 10:50:25  2016-03-28 00:00:00  2016-04-02 14:18:02
49997  2016-04-02 14:44:48  2016-04-02 00:00:00  2016-04-04 11:47:27
49998  2016-03-08 19:25:42  2016-03-08 00:00:00  2016-04-05 16:45:07
49999  2016-03-14 00:42:12  2016-03-13 00:00:00  2016-04-06 21:18:48

[48571 rows x 3 columns]

In [41]:
print("Date Crawled")
print(autos["dateCrawled"].value_counts(normalize=True, dropna=False).sort_index(ascending=True))
print("\n")
print("Ad Created")
print(autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index(ascending=True))
print("\n")
print("Last Seen")
print(autos["lastSeen"].value_counts(normalize=True, dropna=False).sort_index(ascending=True))

Date Crawled
2016-03-05 14:06:30    0.000021
2016-03-05 14:06:40    0.000021
2016-03-05 14:07:04    0.000021
2016-03-05 14:07:08    0.000021
2016-03-05 14:07:21    0.000021
                         ...   
2016-04-07 14:30:09    0.000021
2016-04-07 14:30:26    0.000021
2016-04-07 14:36:44    0.000021
2016-04-07 14:36:55    0.000021
2016-04-07 14:36:56    0.000021
Name: dateCrawled, Length: 46888, dtype: float64


Ad Created
2015-06-11 00:00:00    0.000021
2015-08-10 00:00:00    0.000021
2015-09-09 00:00:00    0.000021
2015-11-10 00:00:00    0.000021
2015-12-05 00:00:00    0.000021
                         ...   
2016-04-03 00:00:00    0.038850
2016-04-04 00:00:00    0.036894
2016-04-05 00:00:00    0.011818
2016-04-06 00:00:00    0.003253
2016-04-07 00:00:00    0.001256
Name: ad_created, Length: 76, dtype: float64


Last Seen
2016-03-05 14:45:46    0.000021
2016-03-05 14:46:02    0.000021
2016-03-05 14:49:34    0.000021
2016-03-05 15:16:11    0.000021
2016-03-05 15:16:47    0.000021
    

This represents 1 month worth of crawled data, however, some of the ads were created several months before. The percentages are very small, because the dates are date/timestamps represented as strings - it looks like each date/timestamp is on its own.

In [42]:
autos["registration_year"].describe()

count    48571.000000
mean      2004.754257
std         88.638548
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The min/max values of registration year are likely placeholders - values such as 1000 and 9999 are not real values. The mean of 2004 makes sense, but it is affected by these outliers.

As a next step, I will remove all vehicles with registration year > 2016 and below 1976 - it is very unlikely that vehicles older than 40 years would be sold.

In [43]:
autos = autos[(autos["registration_year"] >= 1976) & (autos["registration_year"] < 2016)]

In [44]:
autos["registration_year"].describe()

count    45095.000000
mean      2002.862756
std          6.041744
min       1976.000000
25%       1999.000000
50%       2003.000000
75%       2007.000000
max       2015.000000
Name: registration_year, dtype: float64

# Exploring unique brands

In [68]:
unique = autos["brand"].unique()
print(unique)

['peugeot' 'bmw' 'volkswagen' 'smart' 'ford' 'chrysler' 'seat' 'renault'
 'mercedes_benz' 'audi' 'sonstige_autos' 'opel' 'mazda' 'porsche' 'mini'
 'toyota' 'dacia' 'nissan' 'jeep' 'saab' 'volvo' 'mitsubishi' 'jaguar'
 'fiat' 'skoda' 'subaru' 'kia' 'citroen' 'chevrolet' 'hyundai' 'honda'
 'daewoo' 'suzuki' 'trabant' 'land_rover' 'alfa_romeo' 'lada' 'rover'
 'daihatsu' 'lancia']


In [69]:
top_6 = autos['brand'].value_counts()[0:6]

In [70]:
print(top_6.index)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [76]:
price_dict = {}

for c in top_6.index:
    selection = autos.loc[autos["brand"]== c,"price"].mean()
    price_dict[c] = selection

print(price_dict)

{'volkswagen': 5657.430326577759, 'bmw': 8639.826259420865, 'opel': 3001.243288241415, 'mercedes_benz': 8578.830205949656, 'audi': 9346.573158425832, 'ford': 3825.2761050608583}


BMW, Audi and Mercedes have higher mean prices than Opel, Volkswagen and Ford. This does not definitively imply that these brands have higher value vehicles (although this would make sense), because we are not controlling for registration year and type of vehicle.

In [78]:
# creating a new dictionary for mileage #

mileage_dict = {}

for c in top_6.index:
    selection = autos.loc[autos["brand"]== c,"odometer_km"].mean()
    mileage_dict[c] = selection

print(mileage_dict)

{'volkswagen': 128592.35535020477, 'bmw': 132616.02538675128, 'opel': 129500.52029136317, 'mercedes_benz': 131009.1533180778, 'audi': 129196.51866801211, 'ford': 124855.86162716207}


Creating series

In [82]:
price = pd.Series(price_dict)
mileage = pd.Series(mileage_dict)
print(price)
print('\n')
print(mileage)     

volkswagen       5657.430327
bmw              8639.826259
opel             3001.243288
mercedes_benz    8578.830206
audi             9346.573158
ford             3825.276105
dtype: float64


volkswagen       128592.355350
bmw              132616.025387
opel             129500.520291
mercedes_benz    131009.153318
audi             129196.518668
ford             124855.861627
dtype: float64


In [84]:
df = pd.DataFrame(price,columns = ["price"])
print(df)

                     price
volkswagen     5657.430327
bmw            8639.826259
opel           3001.243288
mercedes_benz  8578.830206
audi           9346.573158
ford           3825.276105


In [85]:
df["mileage"] = mileage

In [86]:
print(df)

                     price        mileage
volkswagen     5657.430327  128592.355350
bmw            8639.826259  132616.025387
opel           3001.243288  129500.520291
mercedes_benz  8578.830206  131009.153318
audi           9346.573158  129196.518668
ford           3825.276105  124855.861627


It looks like mileage on the brands is comparable - this means that we have more ground for stating that BMW, Audi and Mercedes' vehicles truly have higher values than vehicles by VW, Opel and Ford.